### PR TITLE
[rostopic] Fix rostopic echo for non rosmsg field

### PR DESCRIPTION
**Pull request by wkentaro**
Close https://github.com/ros/ros_comm/issues/908

### CODE REVIEW

**Comment by dirk-thomas** [comment]
If you run `roslaunch rospy_tutorials talker_listener.launch` and `rostopic echo /chatter --nostr` it prints:

> `data: <string length: 25>`

Therefore I would expect `rostopic echo /chatter --nostr -n 1` to print:

> `<string length: 25>`

but it actually prints:

> `None`

File: tools/rostopic/src/rostopic/__init__.py [hunk 1]
```diff
@@ -1389,6 +1389,12 @@ def eval_fn(m):
 
 def create_value_transform(echo_nostr, echo_noarr):
     def value_transform(val):
+        if not isinstance(val, genpy.Message):
+            if echo_nostr and isinstance(val, str):
+                return None
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: What about this instead:

`return '<string length: %s>' % len(val)`

Comment (wkentaro) [reply tools/rostopic/src/rostopic/__init__.py]: Nice.

File: tools/rostopic/src/rostopic/__init__.py [hunk 2]
```diff
@@ -1389,6 +1389,12 @@ def eval_fn(m):
...
+            elif echo_noarr and isinstance(val, list):
+                return None
```
Comment (dirk-thomas) [inline tools/rostopic/src/rostopic/__init__.py]: Here I don't see how we could print the type information (`<array type: ??? ...>`):

`return '<array length: %s>' % len(val)`

**Comment by wkentaro** [comment]
With 47adb6d, 

```
% rostopic echo /rosout/msg --nostr
<string length: 24>
---

% rostopic echo /rosout/topics --noarr
<array type: string, length: 2>
---
```

**Comment by wkentaro** [comment]
With 9dd94cf, 

**Before**

```
% rostopic echo /rosout --nostr -n1
header:
  seq: 13891
  stamp:
    secs: 1475187530
    nsecs: 501231908
  frame_id: <string length: 0>
level: 2
name: <string length: 9>
msg: <string length: 41>
file: <string length: 11>
function: <string length: 8>
line: 43
topics: <string length: 2>
```

**After**

```
% rostopic echo /rosout --nostr -n1
header:
  seq: 14060
  stamp:
    secs: 1475187547
    nsecs: 301131010
  frame_id: <string length: 0>
level: 2
name: <string length: 9>
msg: <string length: 41>
file: <string length: 11>
function: <string length: 8>
line: 43
topics: <array type: string, length: 2>
---
```

**Comment by dirk-thomas** [comment]
The current patch duplicates a lot of code / logic. I would recommend to create a function like this:

```
def _value_transform(type_information, val, echo_nostr, echo_noarr):
    if echo_noarr and '[' in type_information:
        return '<array type: %s, length: %s>' % \
            (type_information.rstrip('[]'), len(val))
    elif echo_nostr and type_information == 'string':
        return '<string length: %s>' % len(val)
    elif echo_nostr and type_information == 'string[]':
        return '<array type: string, length: %s>' % len(val)
    return None
```

and call it from the two locations accordingly.

**Comment by dirk-thomas** [comment]
This looks good to me. Thank you for providing this patch so fast.

**Comment by mikepurvis** [comment]
Thanks for the quick turnaround, all.

**Pull request merged**